Ready for review: transport and reactive transport

WilkAndy opened the pull request:
ModelGeochemicalDatabase records swap matrix in order to translate bulk compositions back to original basis.  This swap matrix is updated every time a swap is performed by the GeochemistrySpeciesSwapper.  GeochemicalSystem has a method for getting the bulk in the original basis.

ModelGeochemicalDatabase has boolean vectors (basis_species_transported, eqm_species_transported and kin_species_transported) that define whether the species is involved in the transport side of reactive-transport.  GeochemistrySpeciesSwapper swaps these appropriate upon swapping

GeochemicalSystem has methods for computing the bulk-composition mole number of transporting species, both in the current basis and the original basis, based on the current values of molalities

Refs #15693

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

WilkAndy committed:
This is a combination of 40 commits.  The oldest message is at the top, the newest message is down the bottom.

ModelGeochemicalDatabase records swap matrix in order to translate bulk compositions back to original basis.  This swap matrix is updated every time a swap is performed by the GeochemistrySpeciesSwapper.  GeochemicalSystem has a method for getting the bulk in the original basis.

ModelGeochemicalDatabase has boolean vectors (basis_species_transported, eqm_species_transported and kin_species_transported) that define whether the species is involved in the transport side of reactive-transport.  GeochemistrySpeciesSwapper swaps these appropriate upon swapping

GeochemicalSystem has methods for computing the bulk-composition mole number of transporting species, both in the current basis and the original basis, based on the current values of molalities

Refs #15693

GeochemicalSolver now does not get instantiated with a GeochemicalSystem or a ModelGeochemicalDatabase.  Instead all of its methods have the GeochemicalSystem as an argument.  This is so a spatially-varying simulation can have just one GeochemicalSolver that solves the systems at all nodes.  Refs #15693

PertinentGeochemicalSystem now offers the method getIndexOfOriginalBasisSpecies that will be used when retrieving certain information expressed in the original basis

GeochemicalSystem now has a copy-assignment operator that is used in the threadJoin of the Reactors

GeochemicalModelDefinition now provides a const reference to the underlying PertinentGeochemicalSystem.  This is used in the GeochemistryAux to get various things about the original basis species

GeochemicalSystem.h GeochemistryActivityCoefficients.h GeochemistryActivityCoefficientsDebyeHuckel.h GeochemistryIonicStrength.h GeochemistryKineticRateCalculator.h GeochemistrySpeciesSwapper.h PertinentGeochemicalSystem.h now have comparitors in them for testing equality

NearestNodeNumber postprocessor added.  Refs #15749

NearestNodeNumber postprocessor and userobject added.  Refs  #15749

asdf

GeochemistryConsoleOutput now uses the NearestNodeNumber UserObject instead of a point

tests updated to reflect that GeochemistryConsoleOutput now uses NearestNodeNumber UserObject

doc updated regarding GeochemistryConsoleOutput using NearestNodeNumber UserObject

Actions and Reactor UserObjects now do not have getters regarding the point used for output purposes - only the node id is now used.  Also, some enhancements getting ready for spatial dependence

GeochemistyQuantityAux can now extract transported_moles_in_original_basis - useful for reactive transport

spatially-dependent Reactor classes added.  Refs #15693

doco added for spatial dependent sims.  Fixes #15693

overall kinetic mole_additions now allowed in GeochemicalSolver::solveSystem.  This is because these species may be transported in reactive transport and should not be added as basis species, but as kinetic species

mole_additions input to the solver can now include additions of kinetic species so that an external agent can pump kinetic species into the system.

transported mole numbers do not include kinetic species - including them was a mistake because that would mean transport short-circuits any kinetic rates, oh dear.  transporting kinetic species is easy now - just evaluate the kinetic moles and transport them

reorder temperature-setting before flushing in GeochemistryTimeDependentReactor.  This makes it easier to flush with different-temperature fluids.  Also, changes in kinetic-nonminerals in flush.  I think the previous implementation was incorrect, but the Bethke book doesn't actually specify what's supposed to happen for flushing kinetic-nonminerals.  This current implementation makes more sense to me than the previous one

2D GeoTES example added

heat-exchanger mode added to GeochemistryTimeDependentReactor, as well as some refactoring of its code.  Fixes #15768

Actions now have inputs that allow the UserObjects to be block or boundary restrictable.  Fixes #15769

Actions also have a flag that allows a usual moose solve involving Variables and Kernels to proceed.  It defaults to False.  Fixes #15770

PorousFlow DiracKernels now have a coupledVar that multiplies their fluxes.  Fixes #15771

PertinentGeochemicalSystem has a method originalBasisNames so that GeochemistryConsoleOutput can output the bulk composition in the original basis, which i feel might be useful.

Also:

RobPod weber-tensleep geotes example

moved geotes_2D example to combined and golded some tests

computational_efficiencies.md started

Start of Forge simulation

GeochemistrySpatialReactor bug with intel compilers fixed - cannot use size of map while inserting into map

forge equilibrium system

ability to use all relevant minerals in geochemical models.  Refs #15798

kinetic.i changed

forge water chemistry by Stuart suggestions

forge input files and doco

better mesh creation in weber_Tensleep example

forge porous_flow input file test and doco and pics

forge aquifer geochemistry input file and tests

compute efficiencies completed

Ability for GeochemistrySpatialReactor to adaptively change the timestep node-by-node in the case of solver failures.  Refs #15816

remarks on compute efficiencies of fieldsplit method

Reorganize the initialization list in GeochemistrySpatialReactor so that the swaps are implemented in the correct order, ooops

Slightly modify the input Quartz mole number in weber-tensleep so that the porosity is 0.1 not 0.099947
3D pics added to weber-tensleep example

utilize NodalVoidVolume UserObject in examples, now that it is merged.

trying to eliminate CIVET errors

removed png files

loganharbour commented:
Use the draft option :-)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/576725/) on f3a2b97 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15698/site)

This comment will be updated on new commits.

WilkAndy commented:
Hey @cpgr,

You might be interested in the latest commit.  Adaptive timestepping is implemented in the SpatialReactor which might help with convergence.  You still have to set your nodal_void_volume correctly, otherwise you might (will) end up subtracting a species from a node that doesn't have that species.

a

WilkAndy commented:
Hey @cpgr , you can see where your figures fit in at the bottom of

https://mooseframework.inl.gov/docs/PRs/15698/site/modules/geochemistry/tests_and_examples/geotes_weber_tensleep.html

a

cpgr commented:
Great - do you want me to push them here when I've finished or just send them to you?

WilkAndy commented:
I think just email them  as i've got a few more amendments to make too.

a

________________________________
From: Chris Green <notifications@github.com>
Sent: Wednesday, 7 October 2020 3:36 PM
To: idaholab/moose <moose@noreply.github.com>
Cc: Wilkins, Andy (Mineral Resources, Pullenvale) <Andrew.Wilkins@csiro.au>; Assign <assign@noreply.github.com>
Subject: Re: [idaholab/moose] Ready for review: transport and reactive transport (#15698)


Great - do you want me to push them here when I've finished or just send them to you?

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub<https://github.com/idaholab/moose/pull/15698#issuecomment-704704237>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABS6CVP3PD5YKN5NIFX5XJLSJP46TANCNFSM4PT7F5DQ>.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/591281/) on c73f1ee wanted to post the following:

Warning: This PR changes repo size by 5.41 MiB.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/592514/) on 38b56b6 wanted to post the following:

Warning: This PR changes repo size by 5.44 MiB.

lindsayad commented:
Looks like you have a module failure

WilkAndy commented:
> Looks like you have a module failure

Yea, i'm trying to get to the bottom of that.  Only happens for a job with 2 processors running 2 threads on Minimum clang version.   So difficult to me to get to the root problem.  I'm going to mail moose-users.

WilkAndy commented:
Reminding you about this monumental PR, @loganharbour

loganharbour commented:
Looks good to me @WilkAndy.

Will get it in once we can get stuff passing again.

loganharbour commented:
Also: if you keep adding all of this media content (images), I would prefer that you start putting them into large_media.

moosebuild commented:
Job [Non unity build](https://civet.inl.gov/job/592540/) on 38b56b6 : invalidated by @WilkAndy

bison failed to build

WilkAndy commented:
At the moment i don't understand why the spatial_4_threads on Minimum clang version is failing.  I'm still investigating as i can't reproduce the failure on any of my systems.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/615502/) on b950ce5 wanted to post the following:

Warning: This PR changes repo size by 5.40 MiB.

loganharbour commented:
> Warning: This PR changes repo size by 5.40 MiB.

I actually missed this further up. Is it possible for you to move the images into large_media?

WilkAndy commented:
Hi @loganharbour .  So, what i've done is to

1. rebase against next.  This meant a whole lot of trouble because lots of the unit tests failed because of the changes to json in the rest of moose, sigh
2. squashed everything into one giant commit - this just made the next step easier
3. removed the png files because i've added them to large_media.

If everything passes i'll squash (2) and (3) together to properly remove the png files, but i'm using CIVET to check everything, partly because it's more thorough than my testing, and partly because it's a lot faster.

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 77f85490982aa6357b5e9b062ee28694b8dc385d
+Subproject commit e41f7c8f6e0e77b620dbe4a2b24711b89f52599e
```
**Caution!** This contains a submodule update

WilkAndy replied:
Yep, that's deliberate

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 77f85490982aa6357b5e9b062ee28694b8dc385d
+Subproject commit e41f7c8f6e0e77b620dbe4a2b24711b89f52599e
```
**Caution!** This contains a submodule update

WilkAndy commented:
Hey @loganharbour , finally all the tests have passed.  I added the png files to large_media.   I really hope you can merge this - i spent most of the day mucking around on things that had regressed through changes in the framework, which was a bit disheartening!

loganharbour reviewed (approved):
Good to go once next is green again!

WilkAndy commented:
Sounds good, do i have to do anything further, @loganharbour ?

lindsayad commented:
https://civet.inl.gov/job/617745/

lindsayad commented:
I thought that you had decided to skip that test
